crates/key-value-aws/src/store.rs

```diff
@@ -0,0 +1,517 @@
+                .batch_write_item()
+                .set_request_items(request_items)
+                .send()
+                .await
+                .map_err(log_error)?;
+
+            request_items = unprocessed_items.filter(|unprocessed| !unprocessed.is_empty());
+        }
+
+        Ok(())
+    }
+
+    async fn delete_many(&self, keys: Vec<String>) -> Result<(), Error> {
+        let mut data = Vec::with_capacity(keys.len());
+        for key in keys {
+            data.push(
+                WriteRequest::builder()
+                    .delete_request(
+                        DeleteRequest::builder()
+                            .key(PK, AttributeValue::S(key))
+                            .build()
+                            .map_err(log_error)?,
+                    )
+                    .build(),
+            )
+        }
+
+        let mut request_items = Some(HashMap::from_iter([(self.table.to_string(), data)]));
+
+        while request_items.is_some() {
+            let BatchWriteItemOutput {
+                unprocessed_items, ..
+            } = self
+                .client
+                .batch_write_item()
+                .set_request_items(request_items)
+                .send()
+                .await
+                .map_err(log_error)?;
+
+            request_items = unprocessed_items.filter(|unprocessed| !unprocessed.is_empty());
+        }
+
+        Ok(())
+    }
+
+    async fn increment(&self, key: String, delta: i64) -> Result<i64, Error> {
+        let result = self
+            .client
+            .update_item()
+            .table_name(self.table.as_str())
+            .key(PK, AttributeValue::S(key))
+            .update_expression("ADD #val :delta")
+            .expression_attribute_names("#val", VAL)
+            .expression_attribute_values(":delta", AttributeValue::N(delta.to_string()))
+            .return_values(aws_sdk_dynamodb::types::ReturnValue::UpdatedNew)
+            .send()
+            .await
+            .map_err(log_error)?;
+
+        if let Some(updated_attributes) = result.attributes {
+            if let Some(AttributeValue::N(new_value)) = updated_attributes.get(VAL) {
+                return Ok(new_value.parse::<i64>().map_err(log_error))?;
+            }
+        }
+
+        Err(Error::Other("Failed to increment value".into()))
+    }
+
+    async fn new_compare_and_swap(
+        &self,
+        bucket_rep: u32,
+        key: &str,
+    ) -> Result<Arc<dyn spin_factor_key_value::Cas>, Error> {
+        Ok(Arc::new(CompareAndSwap {
+            key: key.to_string(),
+            client: self.client.clone(),
+            table: self.table.clone(),
+            version: Mutex::new(None),
+            bucket_rep,
+        }))
+    }
+}
+
+#[async_trait]
+impl Cas for CompareAndSwap {
+    async fn current(&self) -> Result<Option<Vec<u8>>, Error> {
+        // TransactGetItems fails if concurrent writes are in progress on an item
+        let output = self
```

Comment:
Thinking aloud -- this can be brought closer to ensuring a unique lock: TransactWrite can return the VAL key while setting a lock key -- only under the condition that the lock key doesn't already exist. Combined with deleting the lock key on swap, this could guarantee that only one process can acquire "the lock" assuming all processes call `current` before `swap`. However, this does not prevent another process calling `swap` without first calling `current` and ignoring the lock. But, it may still be worth making this change to get as close as possible to a transaction -- curious to hear thoughts on the best approach for this!

Reply:
Indeed it appears to be, so I will make this change which hopefully drastically simplify the calls in CAS